Solutions/Corvus.Identity.Azure/Corvus/Identity/ClientAuthentication/Azure/Internal/AzureTokenCredentialSourceFromConfiguration.cs

```diff
@@ -0,0 +1,98 @@
+        private readonly IKeyVaultSecretClientFactory secretClientFactory;
+
+        /// <summary>
+        /// Creates an <see cref="AzureTokenCredentialSourceFromConfiguration"/>.
+        /// </summary>
+        /// <param name="secretClientFactory">
+        /// Creates new <see cref="SecretClient"/> instances.
+        /// </param>
+        public AzureTokenCredentialSourceFromConfiguration(
+            IKeyVaultSecretClientFactory secretClientFactory)
+        {
+            this.secretClientFactory = secretClientFactory ?? throw new ArgumentNullException(nameof(SecretClient));
+        }
+
+        /// <inheritdoc/>
+        public async ValueTask<IAzureTokenCredentialSource> CredentialSourceForConfigurationAsync(
+            ClientIdentityConfiguration configuration,
+            CancellationToken cancellationToken)
+        {
+            ClientIdentitySourceTypes? identitySourceType = configuration.IdentitySourceType;
+
+            if (identitySourceType is null)
+            {
+                if (!string.IsNullOrWhiteSpace(configuration.AzureAdAppClientId))
+                {
+                    // TODO: there are other possibilities, and we should also detect conflicting config,
+                    identitySourceType = ClientIdentitySourceTypes.ClientIdAndSecret;
+                }
+            }
+
+            if (identitySourceType == ClientIdentitySourceTypes.ClientIdAndSecret)
+            {
+                string secret;
+                if (!string.IsNullOrWhiteSpace(configuration.AzureAdAppClientSecretPlainText))
+                {
+                    secret = configuration.AzureAdAppClientSecretPlainText!;
+                }
+                else if (configuration.AzureAdAppClientSecretInKeyVault is KeyVaultSecretConfiguration keyVaultConfig)
+                {
+                    TokenCredential keyVaultCredential = keyVaultConfig.VaultClientIdentity is not null
+                        ? await (await this.CredentialSourceForConfigurationAsync(keyVaultConfig.VaultClientIdentity, cancellationToken).ConfigureAwait(false)).GetTokenCredentialAsync(cancellationToken).ConfigureAwait(false)
+                        : new ManagedIdentityCredential();
+                    SecretClient keyVaultSecretClient = this.secretClientFactory.GetSecretClientFor(
+                        keyVaultConfig.VaultName, keyVaultCredential);
+                    Response<KeyVaultSecret> secretResponse = await keyVaultSecretClient.GetSecretAsync(keyVaultConfig.SecretName).ConfigureAwait(false);
+
+                    secret = secretResponse.Value.Value;
+                }
+                else
+                {
+                    throw new ArgumentException(
+                        $"Configuration seems to want Azure AD Client ID and Secret, but not enough information provided",
+                        nameof(configuration));
+                }
+
+                return new AzureTokenCredentialSource(new TestableClientSecretCredential(
```

Comment:
It is not test-specific. It's done this way to enable testing, but we return this in all cases, because there isn't really any good way that I could find to do anything else.. Irritatingly, the way they've implemented this class means that this is the only way you can discover what settings were passed in.